flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericExactlyOnceSink.java

```diff
@@ -0,0 +1,173 @@
+		StreamTaskState taskState = super.snapshotOperatorState(checkpointId, timestamp);
+		saveHandleInState(checkpointId);
+		taskState.setFunctionState(state);
+		return taskState;
+	}
+
+	@Override
+	public void restoreState(StreamTaskState state, long recoveryTimestamp) throws Exception {
+		super.restoreState(state, recoveryTimestamp);
+		this.state = (ExactlyOnceState) state.getFunctionState();
+		out = null;
+	}
+
+	@Override
+	public void notifyOfCompletedCheckpoint(long checkpointId) throws Exception {
+		super.notifyOfCompletedCheckpoint(checkpointId);
+		saveHandleInState(checkpointId);
+
+		synchronized (state.pendingHandles) {
+			Set<Long> pastCheckpointIds = state.pendingHandles.keySet();
+			Set<Long> checkpointsToRemove = new HashSet<>();
+			for (Long pastCheckpointId : pastCheckpointIds) {
+				if (pastCheckpointId <= checkpointId) {
+					List<StateHandle<DataInputView>> handles = state.pendingHandles.get(pastCheckpointId);
+					for (StateHandle<DataInputView> handle : handles) {
+						DataInputView in = handle.getState(getUserCodeClassloader());
+						while (in.readByte() == 1) {
+							IN value = serializer.deserialize(in);
+							sendValue(value);
+						}
+						checkpointsToRemove.add(pastCheckpointId);
+					}
+				}
+			}
+			for (Long toRemove : checkpointsToRemove) {
+				state.pendingHandles.remove(toRemove);
+			}
+		}
+	}
+
+	/**
+	 * Write the given element into the backend.
+	 * @param value value to be written
+	 * @throws Exception
+     */
+	protected abstract void sendValue(IN value) throws Exception;
+
+	@Override
+	public void processElement(StreamRecord<IN> element) throws Exception {
+		IN value = element.getValue();
+		if (serializer == null) {
+			typeInfo = TypeExtractor.getForObject(value);
+			serializer = typeInfo.createSerializer(new ExecutionConfig());
+		}
+		//generate initial operator state
+		if (out == null) {
+			out = getStateBackend().createCheckpointStateOutputView(0, 0);
+		}
+		out.writeByte(1);
+		serializer.serialize(value, out);
+	}
+
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+		//don't do anything, since no are a sink
+	}
+
+	/**
+	 * This state is used to keep a list of all StateHandles (essentially references to past OperatorStates) that were
+	 * used since the last completed checkpoint.
+	 **/
+	public class ExactlyOnceState implements StateHandle<Serializable> {
+		protected HashMap<Long, ArrayList<StateHandle<DataInputView>>> pendingHandles;
+
+		public ExactlyOnceState() {
+			pendingHandles = new HashMap<>();
+		}
+
+		@Override
+		public HashMap<Long, ArrayList<StateHandle<DataInputView>>> getState(ClassLoader userCodeClassLoader) throws Exception {
+			return pendingHandles;
+		}
+
+		@Override
+		public void discardState() throws Exception {
+			pendingHandles = new HashMap<>();
+		}
+
+		@Override
+		public long getStateSize() throws Exception {
+			return 0;
```

Comment:
We should be able to calculate the state as the sum of the state handles stored in `pendingHandles`, shouldn't we?